notes/plutus-report/costing.tex

```diff
@@ -0,0 +1,339 @@
+These parameters will be put into the protocol parameters, so we should have a large amount of flexibility if we need to tweak the cost model later (\cref{req:costing-parameters}).
+
+\paragraph{Choosing the cost model parameters}
+The decoupling of the abstract resource units from the pricing units allows us to use a fairly simple rubric for picking cost model parameters: try and actually follow (be correlated with) real costs!
+
+This ensures that resource costs will follow programmers' intuitions (\cref{req:costing-intuitive}), while giving plenty of flexibility in the final pricing of those resources.
+
+If we later change our minds about the cost model parameters (e.g. because what we consider ``typical hardware'' changes), we can re-calibrate our parameters and submit a protocol parameter change.
+
+\subsubsection{Pricing model}
+The conversion from abstract resources to Ada is done by the ledger.
+We call the system which tells us how to do this conversion the ``pricing model''.
+
+\paragraph{Pricing model strategy}
+The current proposed pricing model simply consists of an Ada price for each of the abstract resources.
+So a budget in terms of Space and Time will be turned into a price in terms of Ada by multiplying the components of the limit by their respective prices.
+
+These prices also obviously function as parameters for the pricing model.
+Hence they will also be put in the protocol parameters, so that pricing can be changed with only a protocol parameter change.
+
+\paragraph{Choosing the pricing model parameters}
+Choosing the pricing model parameters is a complex problem that may have strategic considerations.
+At minimum, the prices will need to be high enough to avoid DoS attacks (\cref{req:costing-dos}).
+
+\subsubsection{Validating scripts}
+Putting the pieces together, what happens when a script is validated is:
+\begin{itemize}
+\item
+  The ledger uses the pricing model parameters to compute the price for the stated resource budget in Ada, and ensures that the transaction has sufficient fees to cover this, and that it does not cause us to exceed the per-block budget (\cref{req:costing-block-budget}).
+\item
+  The ledger runs the script, passing in the stated budget and the cost model parameters.
+\item
+  The evaluator either terminates with success, or stops early if it runs out of budget.
+\end{itemize}
+
+\subsection{Abstract resource units and resource budgets}
+We use two abstract resource units:
+\begin{itemize}
+\item Abstract time/CPU usage (``Time'')
+\item Abstract peak memory usage (``Space'')
+\end{itemize}
+
+Script execution is limited in both Time and Space.
+The overall limit (budget) is therefore a pair of some amount of Time and Space.
+
+Crucially, the interpretation of the Space limit is that it is a limit on \emph{peak} space usage.
+This is because the limit is mainly to prevent crashes due to exceeding a machine's available RAM, not to compensate the slot leader for RAM usage.
```

Comment:
Remember, we have to do this in an abstract way so you get the same answer on every machine. Whether a transaction validates can't depend on how keen your OOM killer is! So we need *some* machine independent way of saying what "using too much memory" means for us.

Reply:
> Whether a transaction validates can't depend on how keen your OOM killer is!

Why not? If your machine often OOMs when it shouldn't, you're not supposed to be a full node -- and that's it. Is it any less specific than "we have a certain memory model in mind and if your hardware is not the one that we think of, then `ExMemory` and actual memory consumption are quite less related to each other"? Instead of being "vaguely precise" about abstract units, we can be "vaguely precise" about actual memory consumption (and we can take stats from GHC RTS, not just defer everything to an external OOM killer).

Anyway, I think it's worth to discuss this with the node people. As the user I would be more satisfied with something like "having 2 GB of RAM should be enough for all cases and if you have less memory or some atypical hardware, then it may or may not work out and you're on your own" than "we can precisely calculate abstract units, but estimating actual memory consumption is hard and if you have atypical hardware, then you may hit an OOM without exhausting the budget". I.e. I think that hand-waving actual memory consumption is better than being precise about abstract units people can't reason about.

Reply:
But then we have no way to check whether a script fits within its expected limits without running it and measuring memory (which is hard), and would happen on a users' machine, which is *not* a typical node operator machine.

We want two things:
- A way to stop people using too much memory...
- ... such that everyone can agree on whether or not you are using too much memory.

Note that we are sort of doing what you suggest: we're starting from actual resource usage, making a model to predict it, and then effectively "freezing" that by decoupling it from the actual resource usage. 

>  think that hand-waving actual memory consumption is better than being precise about abstract units people can't reason about.

What can't they reason about? Having it track programmer intuition is a key requirement. I don't think what we're doing will violate that too much. There may be some cases where we say it uses lots of memory when you don't think it should, but by and large doing things that you expect to make the memory usage better should also help our abstract measure.

Reply:
> But then we have no way to check whether a script fits within its expected limits without running it and measuring memory (which is hard)

Why do we need to measure the exact memory consumption? It it runs, it runs. Otherwise, it OOMs and dies. If scripts that run on other people machines die on yours, it's not a suitable machine. If you are a script writer, you pick a "normal" machine and tweak your scripts to run in substantially less than X megabytes (where X is our recommendation). If you violate that requirement, but scripts still get through the network (if we decide to recommend a bit less than what can actually be processed to have a bit of additional capacity), you consider yourself lucky. And if it doesn't get through the network, it's on you. And we can report "your script runs in that amount of megabytes" simply by printing GHC RTS info (there do exist pitfalls of course, but likely not as many as if you try to count allocations manually).

I'm not suggesting tracking memory usage is worthless, it's just that it seems rather hard and it would be nice if we could focus on time tracking (and other things) at least for V1. But maybe I simply misunderstand the problem at hand.

Reply:
> If scripts that run on other people machines die on yours, it's not a suitable machine.

Who sets the standard? How does this work during block diffusion? Someone with a beefy machine produces a block that a bunch of nodes on the network can't process without an OOM; now the definition of "suitable" has changed arbitrarily. It's *really* important to have as objective a standard of "is this a valid transaction" as possible. We're doing distributed consensus, we don't want to rely on centralized decisions much. We are effectively encoding the "recommended stats" in the form of our cost model and pricing: but these get changed via protocol parameter updates in a way that the network has to agree to. Just letting it be "whatever goes" is going to be chaos, I think.

> If you are a script writer, you pick a "normal" machine and tweak your scripts to run in substantially less than X megabytes (where X is our recommendation).

No, we can't require this. Concievably we could require this of developers, but the resource usage of a script also depends on its arguments, so every user who *uses* the script must also calculate how long it runs for to make sure it's okay. Saying "literally everyone must use a standard machine (and not have contention or anything)" is really not viable.

Reply:
(We can keep discussing this but I'm going to merge this PR for now.)